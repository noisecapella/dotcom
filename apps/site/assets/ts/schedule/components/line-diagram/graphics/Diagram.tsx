import { max } from "lodash";
import React, { ReactElement } from "react";
import { LineDiagramStop } from "../../__schedule";
import {
  hasBranchLines,
  isBranchTerminusStop,
  areOnDifferentBranchLines,
  diagramWidth,
  isMergeStop
} from "../line-diagram-helpers";
import Line from "./Line";
import Merges from "./Merge";
import Stop from "./Stop";
import { routeToModeName } from "../../../../helpers/css";
import { getCurrentState } from "../../../store/ScheduleStore";
import { DirectionId } from "../../../../__v3api";
import { isAGreenLineRoute } from "../../../../models/route";
import { BASE_LINE_WIDTH, DiagonalHatchPattern } from "./graphic-helpers";

interface DiagramProps {
  lineDiagram: LineDiagramStop[];
}

const branchingDescription = (lineDiagram: LineDiagramStop[]): string => {
  const mergeStops = lineDiagram.filter(isMergeStop);
  const branchText = mergeStops.map(mergeStop => {
    const mergeStopName = mergeStop.route_stop.name;
    const branchNames = mergeStop.stop_data
      .filter(sd => sd.branch)
      .map(sd => sd.branch);
    return `Use ${mergeStopName} to transfer to trains on ${branchNames.join(
      " or "
    )} branches. `;
  });
  return branchText.join("");
};
const diagramDescription = (
  lineDiagram: LineDiagramStop[],
  direction: DirectionId
): string => {
  const text = `${lineDiagram.length} stops`;
  const {
    direction_destinations: destinations,
    direction_names: names
  } = lineDiagram[0].route_stop.route!;
  if (destinations) {
    return `${text} going ${names[direction]} to ${destinations[direction]}`;
  }
  return text;
};

const Diagram = ({
  lineDiagram
}: DiagramProps): ReactElement<HTMLElement> | null => {
  const { selectedDirection } = getCurrentState();
  const width = diagramWidth(
    max(lineDiagram.map(ld => ld.stop_data.length)) || 1
  );

  return (
    <>
      <svg
        xmlns="http://www.w3.org/2000/svg"
        role="img"
        aria-labelledby="diagram-title diagram-desc"
        className={`line-diagram-svg ${routeToModeName(
          lineDiagram[0].route_stop.route!
        )}`}
        width={`${width + 4}px`}
        height="100%"
        style={{ left: BASE_LINE_WIDTH / 2 }}
      >
        <title id="diagram-title">
          Line diagram for{" "}
          {isAGreenLineRoute(lineDiagram[0].route_stop.route!)
            ? "Green Line"
            : lineDiagram[0].route_stop.route!.name}
        </title>
        <desc id="diagram-desc">
          {diagramDescription(lineDiagram, selectedDirection)}
          {hasBranchLines(lineDiagram) && branchingDescription(lineDiagram)}
        </desc>

        <defs>{DiagonalHatchPattern}</defs>
        {/* If there are multiple branches, draw the lines and curves for those */
        hasBranchLines(lineDiagram) && <Merges lineDiagram={lineDiagram} />}

        {/* Draw lines between stops */
        lineDiagram.map((current, idx, self) => {
          if (self[idx + 1]) {
            if (
              isBranchTerminusStop(current) &&
              areOnDifferentBranchLines(current, self[idx + 1])
            )
              return null; // don't draw a line from a branch end to the main line

            return (
              <Line
                key={`${current.route_stop.id}-${self[idx + 1].route_stop.id}`}
                from={current}
                to={self[idx + 1]}
              />
            );
          }
          return null; // this is the end of the line, nothing to draw a line to
        })}

        {/* Draw circles for each stop */
        lineDiagram.map(stop => (
          <Stop key={stop.route_stop.id} stop={stop} />
        ))}
      </svg>
    </>
  );
};

export default Diagram;
