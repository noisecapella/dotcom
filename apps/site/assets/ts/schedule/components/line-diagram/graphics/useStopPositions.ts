import { useDispatch } from "react-redux";
import { useRef, useEffect, useCallback } from "react";
import { LineDiagramStop } from "../../__schedule";
import { isMergeStop } from "../line-diagram-helpers";
import { BASE_LINE_WIDTH, BRANCH_SPACING } from "./graphic-helpers";

export interface RefList {
  [key: string]: React.MutableRefObject<HTMLElement | null>;
}

function useStopRef(prev: RefList, stop: LineDiagramStop): RefList {
  return {
    ...prev,
    [stop.route_stop.id]: useRef<HTMLElement>(null)
  };
}

const stopById = (stopId: string): ((stop: LineDiagramStop) => boolean) =>
  // eslint-disable-next-line camelcase
  ({ route_stop }) => stopId === route_stop.id;

const xCoordForStop = (stop: LineDiagramStop): number => {
  if (isMergeStop(stop)) {
    return BASE_LINE_WIDTH + 1;
  }
  // usually just a function of stop_data length
  return BRANCH_SPACING * (stop.stop_data.length - 1) + BASE_LINE_WIDTH + 1;
};

export default function useStopPositions(
  stops: LineDiagramStop[]
): [RefList, () => void] {
  const stopRefsMap: RefList = stops.reduce(useStopRef, {});
  const dispatchStopCoords = useDispatch();
  const updateAllStops = useCallback((): void => {
    Object.entries(stopRefsMap).forEach(([stopId, ref]) => {
      const x = xCoordForStop(stops.find(stopById(stopId))!);
      let coordinates = null;
      if (ref && ref.current) {
        const { offsetTop, offsetHeight } = ref.current;
        const y = offsetTop + offsetHeight / 2;
        coordinates = [x, y];
      }
      dispatchStopCoords({
        type: "set",
        stop: stopId,
        coords: coordinates
      });
    });
  }, [dispatchStopCoords, stopRefsMap, stops]);

  useEffect(() => {
    updateAllStops();
    window.addEventListener("resize", updateAllStops);
    return () => window.removeEventListener("resize", updateAllStops);
  }, [updateAllStops]);

  return [stopRefsMap, updateAllStops];
}
