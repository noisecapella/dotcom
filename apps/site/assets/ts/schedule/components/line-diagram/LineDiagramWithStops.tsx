import React, { ReactElement } from "react";
import Diagram from "./graphics/Diagram";
import useStopPositions, { RefList } from "./graphics/useStopPositions";
import LiveVehicles from "./LiveVehicleIcons";
import { LineDiagramStop, RouteStop } from "../__schedule";
import StopCardList from "./StopCardList";

export const StopRefContext = React.createContext<[RefList, () => void]>([
  {},
  () => {}
]);

const LineDiagramWithStops = (props: {
  stops: LineDiagramStop[];
  predictionUrl: string;
  handleStopClick: (stop: RouteStop) => void;
  channel: string;
}): ReactElement<HTMLElement> => {
  const { stops, handleStopClick, predictionUrl, channel } = props;
  // create a ref for each stop - we will use this to track the location of the stop so we can place the line diagram bubbles
  const [stopRefsMap, updateAllStopCoords] = useStopPositions(stops);

  return (
    <StopRefContext.Provider value={[stopRefsMap, updateAllStopCoords]}>
      <div className="m-schedule-diagram">
        <LiveVehicles channel={channel} />
        <Diagram lineDiagram={stops} />
        <StopCardList
          stops={stops}
          predictionUrl={predictionUrl}
          handleStopClick={handleStopClick}
        />
      </div>
    </StopRefContext.Provider>
  );
};

export default LineDiagramWithStops;
