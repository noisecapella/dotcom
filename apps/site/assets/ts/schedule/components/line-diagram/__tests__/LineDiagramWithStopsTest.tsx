import React, { MutableRefObject } from "react";
import * as redux from "react-redux";
import { mount, ReactWrapper } from "enzyme";
import { cloneDeep, merge } from "lodash";
import { RouteType } from "../../../../__v3api";
import { LineDiagramStop, CrowdingType } from "../../__schedule";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import outwardLineDiagram from "./lineDiagramData/outward.json"; // not a full line diagram
import LineDiagramWithStops from "../LineDiagramWithStops";
import { createLineDiagramCoordStore } from "../graphics/graphic-helpers";
import StopListWithBranches from "../StopListWithBranches";
import * as UseStopPositions from "../graphics/useStopPositions";
import * as simpleLiveData from "./lineDiagramData/live-data.json";
import { LiveDataByStop } from "../__line-diagram";

const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];
let lineDiagramBranchingOut = (outwardLineDiagram as unknown) as LineDiagramStop[];

const route = {
  type: 3 as RouteType,
  name: "route 1",
  long_name: "route 1 long name",
  color: "F00B42",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
};

lineDiagram.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

lineDiagramBranchingOut.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

let lineDiagramBranchingIn = cloneDeep(lineDiagramBranchingOut).reverse();
const CRroute = merge(cloneDeep(route), { type: 2 as RouteType });
lineDiagramBranchingIn.forEach(({ route_stop }) => {
  route_stop.route = CRroute;
  if (route_stop["is_terminus?"]) {
    route_stop["is_beginning?"] = !route_stop["is_beginning?"];
  }
});

const handleStopClick = () => {};
const liveData = (simpleLiveData as unknown) as LiveDataByStop;
const liveDataWithCrowding = (cloneDeep(
  simpleLiveData
) as unknown) as LiveDataByStop;
(liveDataWithCrowding["line-stop2"][0].time_data_with_crowding_list[0]
  .crowding as CrowdingType) = "not_crowded";
const store = createLineDiagramCoordStore(lineDiagram);
const spy = jest.spyOn(UseStopPositions, "default");

// mock the redux state so that snapshot has positioned stops
const mockState = lineDiagram.reduce(
  (acc, stop, index) => ({
    ...acc,
    [stop.route_stop.id]: [10, index * 20 + 30]
  }),
  {}
);
jest
  .spyOn(redux, "useSelector")
  .mockImplementation(selector => selector(mockState));

describe("LineDiagramWithStops", () => {
  let wrapper: ReactWrapper;
  beforeEach(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <LineDiagramWithStops
          stops={lineDiagram}
          handleStopClick={handleStopClick}
          liveData={liveData}
        />
      </redux.Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("uses the useStopPositions hook", () => {
    expect(spy).toHaveBeenCalled();
  });

  it("shows <StopListWithBranches /> if the line has branches", () => {
    // wrapper's line diagram has no branches
    expect(wrapper.find(StopListWithBranches)).toHaveLength(0);
    const wrapperWithBranches = mount(
      <redux.Provider store={store}>
        <LineDiagramWithStops
          stops={lineDiagramBranchingOut}
          handleStopClick={handleStopClick}
          liveData={liveData}
        />
      </redux.Provider>
    );
    expect(wrapperWithBranches.find(StopListWithBranches)).toHaveLength(1);
  });

  it("toggles u-no-crowding-data class if crowding present", () => {
    expect(wrapper.exists(".u-no-crowding-data")).toBeTruthy();
    const wrapperWithCrowding = mount(
      <redux.Provider store={store}>
        <LineDiagramWithStops
          stops={lineDiagram}
          handleStopClick={handleStopClick}
          liveData={liveDataWithCrowding}
        />
      </redux.Provider>
    );
    expect(wrapperWithCrowding.exists(".u-no-crowding-data")).toBeFalsy();
  });
});
