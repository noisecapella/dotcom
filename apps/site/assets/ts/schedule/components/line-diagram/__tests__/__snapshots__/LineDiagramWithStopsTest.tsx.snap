// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineDiagramWithStops renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <LineDiagramWithStops stops={{...}} handleStopClick={[Function: handleStopClick]} predictionUrl=\\"\\" channel=\\"\\">
    <div className=\\"m-schedule-diagram\\">
      <LiveVehicles channel=\\"\\" />
      <Diagram lineDiagram={{...}}>
        <svg xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" aria-labelledby=\\"diagram-title diagram-desc\\" className=\\"line-diagram-svg bus\\" width=\\"40px\\" height=\\"100%\\" style={{...}}>
          <title id=\\"diagram-title\\">
            Line diagram for
             
            route 1
          </title>
          <desc id=\\"diagram-desc\\">
            4 stops going Outbound to Begin
          </desc>
          <defs>
            <pattern id=\\"diagonalHatch\\" width=\\"10\\" height=\\"10\\" patternTransform=\\"rotate(45 0 0)\\" patternUnits=\\"userSpaceOnUse\\">
              <line className=\\"line-diagram-svg__line\\" y2={8} strokeWidth=\\"16px\\" />
            </pattern>
          </defs>
          <Line from={{...}} to={{...}}>
            <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"10px\\" x1=\\"11px\\" x2=\\"11px\\" y1=\\"30px\\" y2=\\"50px\\" stroke=\\"url(#diagonalHatch)\\" />
          </Line>
          <Line from={{...}} to={{...}}>
            <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"10px\\" x1=\\"11px\\" x2=\\"11px\\" y1=\\"50px\\" y2=\\"70px\\" />
          </Line>
          <Line from={{...}} to={{...}}>
            <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"10px\\" x1=\\"11px\\" x2=\\"11px\\" y1=\\"70px\\" y2=\\"90px\\" />
          </Line>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"30px\\" />
          </Stop>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"50px\\" />
          </Stop>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"70px\\" />
          </Stop>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"90px\\" />
          </Stop>
        </svg>
      </Diagram>
      <StopCardList stops={{...}} predictionUrl=\\"\\" handleStopClick={[Function: handleStopClick]}>
        <ol className=\\"u-no-crowding-data\\">
          <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveHeadsigns={[undefined]}>
            <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
              <section className=\\"m-schedule-diagram__content\\">
                <header className=\\"m-schedule-diagram__stop-heading\\">
                  <h4 className=\\"m-schedule-diagram__stop-link\\">
                    <a href=\\"/stops/line-origin\\">
                      <MatchHighlight text=\\"Origin\\" matchQuery={[undefined]}>
                        <span>
                          Origin
                        </span>
                      </MatchHighlight>
                    </a>
                  </h4>
                  <div className=\\"m-schedule-diagram__features\\">
                    <TooltipWrapper tooltipText=\\"Parking\\" tooltipOptions={{...}}>
                      <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Parking\\" title=\\"Parking\\">
                        <span className=\\"notranslate c-svg__icon-parking-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      </span>
                    </TooltipWrapper>
                  </div>
                </header>
                <div className=\\"m-schedule-diagram__stop-details\\">
                  <div className=\\"m-schedule-diagram__connections\\" />
                  <div className=\\"m-schedule-diagram__alert\\">
                    Detour
                  </div>
                </div>
                <footer className=\\"m-schedule-diagram__footer\\">
                  <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                    View schedule
                  </button>
                </footer>
              </section>
            </li>
          </StopCard>
          <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveHeadsigns={[undefined]}>
            <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
              <section className=\\"m-schedule-diagram__content\\">
                <header className=\\"m-schedule-diagram__stop-heading\\">
                  <h4 className=\\"m-schedule-diagram__stop-link\\">
                    <a href=\\"/stops/line-stop1\\">
                      <span className=\\"notranslate c-svg__icon-alerts-triangle\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      <span className=\\"sr-only\\">
                        Service alert or delay
                      </span>
                       
                      <MatchHighlight text=\\"Stop 1\\" matchQuery={[undefined]}>
                        <span>
                          Stop 1
                        </span>
                      </MatchHighlight>
                    </a>
                  </h4>
                  <div className=\\"m-schedule-diagram__features\\" />
                </header>
                <div className=\\"m-schedule-diagram__stop-details\\">
                  <div className=\\"m-schedule-diagram__connections\\">
                    <TooltipWrapper href=\\"/schedules/Orange/line\\" tooltipText=\\"Orange Line\\" tooltipOptions={{...}}>
                      <a href=\\"/schedules/Orange/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Orange Line\\" title=\\"Orange Line\\">
                        <span className=\\"m-schedule-diagram__connection\\">
                          <span className=\\"notranslate c-svg__icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                        </span>
                      </a>
                    </TooltipWrapper>
                    <TooltipWrapper href=\\"/schedules/Green-C/line\\" tooltipText=\\"Green Line C\\" tooltipOptions={{...}}>
                      <a href=\\"/schedules/Green-C/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Green Line C\\" title=\\"Green Line C\\">
                        <span className=\\"m-schedule-diagram__connection\\">
                          <span className=\\"notranslate c-svg__icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                        </span>
                      </a>
                    </TooltipWrapper>
                  </div>
                </div>
                <footer className=\\"m-schedule-diagram__footer\\">
                  <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                    View schedule
                  </button>
                </footer>
              </section>
            </li>
          </StopCard>
          <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveHeadsigns={[undefined]}>
            <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
              <section className=\\"m-schedule-diagram__content\\">
                <header className=\\"m-schedule-diagram__stop-heading\\">
                  <h4 className=\\"m-schedule-diagram__stop-link\\">
                    <a href=\\"/stops/line-stop2\\">
                      <span className=\\"notranslate c-svg__icon-alerts-triangle\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      <span className=\\"sr-only\\">
                        Service alert or delay
                      </span>
                       
                      <MatchHighlight text=\\"Stop 2\\" matchQuery={[undefined]}>
                        <span>
                          Stop 2
                        </span>
                      </MatchHighlight>
                    </a>
                  </h4>
                  <div className=\\"m-schedule-diagram__features\\">
                    <TooltipWrapper tooltipText=\\"Accessible\\" tooltipOptions={{...}}>
                      <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Accessible\\" title=\\"Accessible\\">
                        <span className=\\"notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      </span>
                    </TooltipWrapper>
                  </div>
                </header>
                <div className=\\"m-schedule-diagram__stop-details\\">
                  <div className=\\"m-schedule-diagram__connections\\">
                    <TooltipWrapper href=\\"/schedules/62/line\\" tooltipText=\\"Route 62\\" tooltipOptions={{...}}>
                      <a href=\\"/schedules/62/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Route 62\\" title=\\"Route 62\\">
                        <span className=\\"c-icon__bus-pill--small m-schedule-diagram__connection u-bg--bus\\">
                          62
                        </span>
                      </a>
                    </TooltipWrapper>
                    <TooltipWrapper href=\\"/schedules/67/line\\" tooltipText=\\"Route 67\\" tooltipOptions={{...}}>
                      <a href=\\"/schedules/67/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Route 67\\" title=\\"Route 67\\">
                        <span className=\\"c-icon__bus-pill--small m-schedule-diagram__connection u-bg--bus\\">
                          67
                        </span>
                      </a>
                    </TooltipWrapper>
                  </div>
                </div>
                <footer className=\\"m-schedule-diagram__footer\\">
                  <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                    View schedule
                  </button>
                </footer>
              </section>
            </li>
          </StopCard>
          <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveHeadsigns={[undefined]}>
            <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
              <section className=\\"m-schedule-diagram__content\\">
                <header className=\\"m-schedule-diagram__stop-heading\\">
                  <h4 className=\\"m-schedule-diagram__stop-link\\">
                    <a href=\\"/stops/line-destination\\">
                      <MatchHighlight text=\\"Destination\\" matchQuery={[undefined]}>
                        <span>
                          Destination
                        </span>
                      </MatchHighlight>
                    </a>
                  </h4>
                  <div className=\\"m-schedule-diagram__features\\">
                    <TooltipWrapper tooltipText=\\"Parking\\" tooltipOptions={{...}}>
                      <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Parking\\" title=\\"Parking\\">
                        <span className=\\"notranslate c-svg__icon-parking-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      </span>
                    </TooltipWrapper>
                    <TooltipWrapper tooltipText=\\"Accessible\\" tooltipOptions={{...}}>
                      <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Accessible\\" title=\\"Accessible\\">
                        <span className=\\"notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      </span>
                    </TooltipWrapper>
                  </div>
                </header>
                <div className=\\"m-schedule-diagram__stop-details\\">
                  <div className=\\"m-schedule-diagram__connections\\">
                    <TooltipWrapper href=\\"/schedules/BOAT-A/line\\" tooltipText=\\"Atlantis\\" tooltipOptions={{...}}>
                      <a href=\\"/schedules/BOAT-A/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Atlantis\\" title=\\"Atlantis\\">
                        <span className=\\"m-schedule-diagram__connection\\">
                          <span className=\\"notranslate c-svg__icon-bus-small\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                        </span>
                      </a>
                    </TooltipWrapper>
                  </div>
                </div>
                <footer className=\\"m-schedule-diagram__footer\\">
                  <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                    View schedule
                  </button>
                </footer>
              </section>
            </li>
          </StopCard>
        </ol>
      </StopCardList>
    </div>
  </LineDiagramWithStops>
</Provider>"
`;
